src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt

```diff
@@ -671,20 +675,146 @@ data class ImportContext private constructor(
             attributes = path.stdlibAttributes,
             namespaceFilter = path.namespaceFilter(isCompletion)
         )
+
+        fun from(project: Project, element: RsElement): ImportContext = ImportContext(
+            project = project,
+            mod = element.containingMod,
+            superMods = LinkedHashSet(element.containingMod.superMods),
+            scope = RsWithMacrosScope(project, GlobalSearchScope.allScope(project)),
+            pathParsingMode = PathParsingMode.NO_COLONS,
+            attributes = element.stdlibAttributes,
+            namespaceFilter = { true }
+        )
     }
 }
 
-private val RsPath.pathParsingMode: PathParsingMode get() = when (parent) {
-    is RsPathExpr,
-    is RsStructLiteral,
-    is RsPatStruct,
-    is RsPatTupleStruct -> PathParsingMode.COLONS
-    else -> PathParsingMode.NO_COLONS
+/**
+ * @param toImport  Set of unresolved items that can be imported
+ * @param toQualify Set of unresolved items that can't be imported
+ */
+data class TypeReferencesInfo(
+    val toImport: Set<RsQualifiedNamedElement>,
+    val toQualify: Set<RsQualifiedNamedElement>
+)
+
+/**
+ * Traverses type references in `elements` and collects all items that unresolved in current context.
+ */
+fun RsElement.getTypeReferencesInfoFromElements(
+    elements: Collection<RsElement>,
+    subst: Substitution = emptySubstitution
+): TypeReferencesInfo = getTypeReferencesInfo(this, elements, subst, ::collectImportSubjectsFromTypeReferences)
+
+/**
+ * Traverses types in `ty` and collects all items that unresolved in current context.
+ */
+fun RsElement.getTypeReferencesInfoFromTy(
+    ty: Ty,
+    subst: Substitution = emptySubstitution
+): TypeReferencesInfo = getTypeReferencesInfo(this, listOf(ty), subst, ::collectImportSubjectsFromTy)
+
+private fun <T> getTypeReferencesInfo(
+    context: RsElement,
+    elements: Collection<T>,
+    subst: Substitution,
+    collector: (T, Substitution, MutableSet<RsQualifiedNamedElement>) -> Unit
+): TypeReferencesInfo {
+    val result = hashSetOf<RsQualifiedNamedElement>()
+    elements.forEach { collector(it, subst, result) }
+    return processRawImportSubjects(context, result)
+}
+
+private fun collectImportSubjectsFromTypeReferences(
+    context: RsElement,
+    subst: Substitution,
+    result: MutableSet<RsQualifiedNamedElement>
+) {
+    context.accept(object : RsVisitor() {
+        override fun visitTypeReference(reference: RsTypeReference) =
+            collectImportSubjectsFromTy(reference.type, subst, result)
+
+        override fun visitElement(element: RsElement) =
+            element.acceptChildren(this)
+    })
 }
+
+private fun collectImportSubjectsFromTy(
+    ty: Ty,
+    subst: Substitution,
+    result: MutableSet<RsQualifiedNamedElement>
+) {
+    ty.substitute(subst).visitWith(object : TypeVisitor {
+        override fun visitTy(ty: Ty): Boolean {
+            when (ty) {
+                is TyAdt ->
+                    result.add(ty.item)
```

Comment:
This code doesn't take into account type aliases. It should do it at least for `SpecifyTypeExplicitlyIntention` because I've just merged #4403
For example, 
```rust
use crate::foo::bar_baz;

mod foo {
    pub struct Bar<T>(T);
    pub struct Baz;
    pub type BarBaz = Bar<Baz>;

    pub fn bar_baz() -> BarBaz { Bar(Baz) }
}

fn main() {
    let bar/*caret*/ = bar_baz();
}
```